### PR TITLE
 ENH: linearly spaced date_range (GH 20808)

### LINKED ISSUE #20808: ENH: date_range not working as it intuitively should when specifying start, end, and periods
#### Code Sample, a copy-pastable example if possible

```python
>>> start = pd.Timestamp('2008-01-02 07:51:37.999477')
>>> end = start + pd.Timedelta('2 hours')
>>> pd.date_range(start, end, periods=1000)    # Intuitively a linearly spaced time series
Traceback (most recent call last):

  File "<ipython-input-69-2304a28824c6>", line 1, in <module>
    pd.date_range(start, end, periods=1000)

  File "E:\Anaconda3\lib\site-packages\pandas\core\indexes\datetimes.py", line 2057, in date_range
    closed=closed, **kwargs)

  File "E:\Anaconda3\lib\site-packages\pandas\util\_decorators.py", line 118, in wrapper
    return func(*args, **kwargs)

  File "E:\Anaconda3\lib\site-packages\pandas\core\indexes\datetimes.py", line 324, in __new__
    ambiguous=ambiguous)

  File "E:\Anaconda3\lib\site-packages\pandas\core\indexes\datetimes.py", line 421, in _generate
    raise ValueError('Of the three parameters: start, end, and '

ValueError: Of the three parameters: start, end, and periods, exactly two must be specified
```
#### Problem description

I need a DatetimeIndex object to later use as index in a Series. DatetimeIndex should start at `start`, end at `end` and have a fixed number of elements (1000). Intuitively, this should work with `pd.date_range`, but it doesn't, and I haven't found a good explanation about why this is the case.
I have found a workaround on Stackoverflow (https://stackoverflow.com/questions/25796030/how-can-i-use-pandas-date-range-to-obtain-a-time-series-with-n-specified-perio) that does work:
```python
>>> start = pd.Timestamp('2008-01-02 07:51:37.999477')
>>> end = start + pd.Timedelta('2 hours')
>>> pd.to_datetime(np.linspace(start.value, end.value, 1000))
DatetimeIndex(['2008-01-02 07:51:37.999476992',
               '2008-01-02 07:51:45.206684160',
               '2008-01-02 07:51:52.413891328',
               '2008-01-02 07:51:59.621098496',
               '2008-01-02 07:52:06.828305920',
               '2008-01-02 07:52:14.035513088',
               '2008-01-02 07:52:21.242720256',
               '2008-01-02 07:52:28.449927424',
               '2008-01-02 07:52:35.657134592',
               '2008-01-02 07:52:42.864341760',
               ...
               '2008-01-02 09:50:33.134612224',
               '2008-01-02 09:50:40.341819392',
               '2008-01-02 09:50:47.549026560',
               '2008-01-02 09:50:54.756233728',
               '2008-01-02 09:51:01.963440896',
               '2008-01-02 09:51:09.170648064',
               '2008-01-02 09:51:16.377855488',
               '2008-01-02 09:51:23.585062656',
               '2008-01-02 09:51:30.792269824',
               '2008-01-02 09:51:37.999476992'],
              dtype='datetime64[ns]', length=1000, freq=None)
```

#### Expected Output
```python
>>> start = pd.Timestamp('2008-01-02 07:51:37.999477')
>>> end = start + pd.Timedelta('2 hours')
>>> pd.date_range(start, end, periods=1000)
DatetimeIndex(['2008-01-02 07:51:37.999476992',
               '2008-01-02 07:51:45.206684160',
               '2008-01-02 07:51:52.413891328',
               '2008-01-02 07:51:59.621098496',
               '2008-01-02 07:52:06.828305920',
               '2008-01-02 07:52:14.035513088',
               '2008-01-02 07:52:21.242720256',
               '2008-01-02 07:52:28.449927424',
               '2008-01-02 07:52:35.657134592',
               '2008-01-02 07:52:42.864341760',
               ...
               '2008-01-02 09:50:33.134612224',
               '2008-01-02 09:50:40.341819392',
               '2008-01-02 09:50:47.549026560',
               '2008-01-02 09:50:54.756233728',
               '2008-01-02 09:51:01.963440896',
               '2008-01-02 09:51:09.170648064',
               '2008-01-02 09:51:16.377855488',
               '2008-01-02 09:51:23.585062656',
               '2008-01-02 09:51:30.792269824',
               '2008-01-02 09:51:37.999476992'],
              dtype='datetime64[ns]', length=1000, freq=None)
```

#### Output of ``pd.show_versions()``

<details>

INSTALLED VERSIONS
------------------
commit: None
python: 3.6.4.final.0
python-bits: 64
OS: Windows
OS-release: 7
machine: AMD64
processor: Intel64 Family 6 Model 44 Stepping 2, GenuineIntel
byteorder: little
LC_ALL: None
LANG: en
LOCALE: None.None

pandas: 0.22.0
pytest: 3.3.2
pip: 9.0.1
setuptools: 38.4.0
Cython: 0.27.3
numpy: 1.14.2
scipy: 1.0.1
pyarrow: None
xarray: None
IPython: 6.2.1
sphinx: 1.6.6
patsy: 0.5.0
dateutil: 2.6.1
pytz: 2017.3
blosc: None
bottleneck: 1.2.1
tables: 3.4.2
numexpr: 2.6.4
feather: None
matplotlib: 2.1.2
openpyxl: 2.4.10
xlrd: 1.1.0
xlwt: 1.3.0
xlsxwriter: 1.0.2
lxml: 4.1.1
bs4: 4.6.0
html5lib: 0.9999999
sqlalchemy: 1.2.1
pymysql: None
psycopg2: None
jinja2: 2.10
s3fs: None
fastparquet: None
pandas_gbq: None
pandas_datareader: None

</details>

### PR DESCRIPTION
**Pull request by onnoeberhard**
- [X] closes #20808
- [X] tests added / passed
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

### CODE REVIEW

File: pandas/tests/indexes/datetimes/test_date_range.py [hunk 1]
```diff
@@ -162,6 +162,21 @@ def test_date_range_ambiguous_arguments(self):
         with tm.assert_raises_regex(ValueError, msg):
             date_range(start, end, periods=10, freq='s')
 
+    def test_date_range_convenience_periods(self):
```
Comment (mroeschke) [inline pandas/tests/indexes/datetimes/test_date_range.py]: Could you also test this with the `tz` arg specified? Would also be good to test a `tz` where there is a day light savings transition between `start` and `end`.

File: doc/source/whatsnew/v0.23.0.txt [hunk 2]
```diff
@@ -450,6 +450,7 @@ Other Enhancements
 - Updated ``to_gbq`` and ``read_gbq`` signature and documentation to reflect changes from
   the Pandas-GBQ library version 0.4.0. Adds intersphinx mapping to Pandas-GBQ
   library. (:issue:`20564`)
+- :func:`pandas.core.indexes.datetimes.date_range` now returns a linearly spaced DatetimeIndex if ``start``, ``stop``, and ``periods`` are specified, but ``freq`` is not. (:issue:`20808`)
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: - func:\`date_range\` should suffice here
- DatetimeIndex --> \`\`DatetimeIndex\`\`

File: pandas/core/indexes/datetimes.py [hunk 3]
```diff
@@ -2704,7 +2715,36 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
...
+    `periods` are declared.
+
+    >>> date_range('2018-04-24', '2018-04-27', periods=3, box=False)
+    array(['2018-04-24T00:00:00.000000000', '2018-04-25T12:00:00.000000000',
+           '2018-04-27T00:00:00.000000000'], dtype='datetime64[ns]')
     """
+
+    # See https://github.com/pandas-dev/pandas/issues/20808
+    if freq is None and periods is not None and start is not None \
+            and end is not None:
```
Comment (jschendel) [inline pandas/core/indexes/datetimes.py]: A little bit cleaner to use `if freq is None and com._all_not_none(periods, start, end):`

File: pandas/core/indexes/datetimes.py [hunk 4]
```diff
@@ -2704,7 +2715,36 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
...
+        if is_float(periods):
+            periods = int(periods)
+        elif not is_integer(periods):
+            msg = 'periods must be a number, got {periods}'
+            raise TypeError(msg.format(periods=periods))
+
+        start = Timestamp(start)
+        end = Timestamp(end)
```
Comment (jschendel) [inline pandas/core/indexes/datetimes.py]: `start` and `end` should be normalized if `normalize=True` is passed to `date_range`

**Comment by jschendel** [comment]
For consistency, it seems like this behavior should also be implemented for `bdate_range`, `timedelta_range`, and `interval_range`.  Doesn't really make sense for `period_range`, since a freq is always required there, and can probably skip `cdate_range` since it's been deprecated (although not removed).

**Comment by onnoeberhard** [comment]
Alright, thank you for all the feedback. I have implemented almost everything mentioned, except for the implementation of a similar feature in the functions `bdate_range`, `timedelta_range` and `interval_range`, because I have not yet used these functions and don't know how to properly implement it in an intuitive way.

File: pandas/core/indexes/datetimes.py [hunk 5]
```diff
@@ -2704,7 +2715,49 @@ def date_range(start=None, end=None, periods=None, freq='D', tz=None,
...
+    function that is used when all three parameters `start`, `end`, and
+    `periods` are declared. If this results in anything else than a
+    DatetimeIndex (like in this example), you cannot specify `tz` or `name`.
+
+    >>> date_range('2018-04-24', '2018-04-27', periods=3, box=False)
+    array(['2018-04-24T00:00:00.000000000', '2018-04-25T12:00:00.000000000',
+           '2018-04-27T00:00:00.000000000'], dtype='datetime64[ns]')
     """
+
+    # See https://github.com/pandas-dev/pandas/issues/20808
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: this needs to go in the DTI constructor itself. we already do some of this validation, it needs to fit in there. Further you don't need to worry about lots of other things that you are repeating, e.g. tz, which are already handled.

**Review by jreback — CHANGES_REQUESTED** [review]
also would be nice to add this to TDI

**Comment by onnoeberhard** [comment]
I have moved everything to the DTI constructor. With that I have removed the functionality to declare kwargs to `to_datetime`, but I think it is much cleaner overall like this.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thank @onnoeberhard

**Pull request merged**